examples/chat.rs

```diff
@@ -0,0 +1,96 @@
+//! A chat server that broadcasts a message to all connections.
+
+extern crate tokio_core;
+extern crate futures;
+
+use std::collections::HashMap;
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::iter;
+use std::env;
+use std::io::BufReader;
+
+use tokio_core::net::TcpListener;
+use tokio_core::reactor::Core;
+use tokio_core::io::{self, Io};
+
+use futures::stream::{self, Stream};
+use futures::Future;
+
+fn main() {
+    let addr = env::args().nth(1).unwrap_or("127.0.0.1:8080".to_string());
+    let addr = addr.parse().unwrap();
+    // We are single-threaded, so we can just use Rc and RefCell.
+    let connections = Rc::new(RefCell::new(HashMap::new()));
+
+    let mut core = Core::new().unwrap();
+    let handle = core.handle();
+    let socket = TcpListener::bind(&addr, &handle).unwrap();
+    println!("Listening on: {}", addr);
+
+    let connections = connections.clone();
+    let future = socket.incoming().for_each(move |(stream, addr)| {
+        let connections = connections.clone();
+        let handle_inner = handle.clone();
```

Comment:
It's ok to defer these clones until they're needed as opposed to happening immediately.